### PR TITLE
fix getmodel()

### LINKED ISSUE #2098: Due to FP16, model returned from DistriOptimizer.optimize() has different accuracy compared to cached model.
The test result in the Training of lenet5:
```
2017-12-25 13:28:46 INFO  DistriOptimizer$:682 - [Wall Clock 74.458101819s] Validate model...
2017-12-25 13:28:47 INFO  DistriOptimizer$:724 - Top1Accuracy is Accuracy(correct: 9773, count: 10000, accuracy: 0.9773)
2017-12-25 13:28:47 INFO  DistriOptimizer$:724 - Top5Accuracy is Accuracy(correct: 10000, count: 10000, accuracy: 1.0)
2017-12-25 13:28:47 INFO  DistriOptimizer$:724 - Loss is (Loss: 11.180599, count: 160, Average Loss: 0.06987874)
evaluate: 0.978
validation: 0.9773
```
Top1Accuracy during validation is 97.73%.
Result of `Optimizer.optimize().evaluate(validationSet, Array(new Top1Accuracy[Float]()), Some(256))` is 97.8%. This result maybe lower or higher than Top1Accuracy during validation.

### PR DESCRIPTION
**Pull request by qiuxin2012**
## What changes were proposed in this pull request?
get weights from cached weights instead of AllreduceParameter.
Sync weights once weights is computed by OptimMethod.

## How was this patch tested?
unit tests, manual tests

## Related links or issues (optional)
fixed #2098

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala [hunk 1]
```diff
@@ -747,24 +747,29 @@ object DistriOptimizer {
...
+      val length = taskSize + (if (pid < extraSize) 1 else 0)
+      Iterator.single((
+        Map(pid -> cached.modelWeights(0).narrow(1, start + 1, length).clone()),
+        Map(pid -> cached.modelGradients(0).narrow(1, start + 1, length).clone())))
     }).reduce((a, b) => (a._1 ++ b._1, a._2 ++ b._2))
 
     val parameterArray = trainingModel.parameters()
     (0 until parameterArray._2.length).foreach(i =>
       parameterArray._2(i).resizeAs(parameterArray._1(i))
     )
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: delete unuse

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala [hunk 2]
```diff
@@ -747,24 +747,29 @@ object DistriOptimizer {
     parameters: AllReduceParameter[T],
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: unuse

**Comment by zhangxiaoli73** [comment]
LGTM

**Comment by qiuxin2012** [comment]
jenkins passed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala [hunk 3]
```diff
@@ -360,6 +354,12 @@ object DistriOptimizer {
           driverMetrics.add("compute weight average", System.nanoTime() - time)
           time = System.nanoTime()
           parameters.sendWeightPartition()
+          /*
+            Note: All models in `cached` share the same storage for weights, so we only need to
+            copy the weights from parameter server into the first model's weights.
+           */
+          val weightsResult = parameters.getWeights(modelCache.modelWeights.head)
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: Couldn't sync here. Find other solution.

**Pull request closed without merging**